src/EventStore.Core/Services/Storage/EpochManager/EpochManager.cs

```diff
@@ -177,101 +213,139 @@ public class EpochManager : IEpochManager {
+				throw new ArgumentException(
+					$"Cannot add Epoch {epochNumber}, new Epoch numbers must be greater than the Last Epoch  {LastEpochNumber}.",
+					nameof(epochNumber));
+			}
+
+			var epoch = WriteEpochRecordWithRetry(epochNumber, Guid.NewGuid(),
+				_lastCachedEpoch?.Value.EpochPosition ?? -1, _instanceId);
+			AddEpochToCache(epoch);
 		}
 
-		private EpochRecord WriteEpochRecordWithRetry(int epochNumber, Guid epochId, long lastEpochPosition, Guid instanceId) {
+		private EpochRecord WriteEpochRecordWithRetry(int epochNumber, Guid epochId, long lastEpochPosition,
+			Guid instanceId) {
 			long pos = _writer.Checkpoint.ReadNonFlushed();
 			var epoch = new EpochRecord(pos, epochNumber, epochId, lastEpochPosition, DateTime.UtcNow, instanceId);
 			var rec = new SystemLogRecord(epoch.EpochPosition, epoch.TimeStamp, SystemRecordType.Epoch,
 				SystemRecordSerialization.Json, epoch.AsSerialized());
-
+			Log.Debug(
+							"=== Writing E{epochNumber}@{epochPosition}:{epochId:B} (previous epoch at {lastEpochPosition}). L={leaderId:B}.",
+							epochNumber, epoch.EpochPosition, epochId, lastEpochPosition, epoch.LeaderInstanceId);
 			if (!_writer.Write(rec, out pos)) {
 				epoch = new EpochRecord(pos, epochNumber, epochId, lastEpochPosition, DateTime.UtcNow, instanceId);
 				rec = new SystemLogRecord(epoch.EpochPosition, epoch.TimeStamp, SystemRecordType.Epoch,
 					SystemRecordSerialization.Json, epoch.AsSerialized());
 				if (!_writer.Write(rec, out pos))
-					throw new Exception(string.Format("Second write try failed at {0}.", epoch.EpochPosition));
+					throw new Exception($"Second write try failed at {epoch.EpochPosition}.");
 			}
-
-			Log.Debug("=== Writing E{epochNumber}@{epochPosition}:{epochId:B} (previous epoch at {lastEpochPosition}). L={leaderId:B}.",
-				epochNumber, epoch.EpochPosition, epochId, lastEpochPosition, epoch.LeaderInstanceId);
-
+			_writer.Flush();
 			_bus.Publish(new SystemMessage.EpochWritten(epoch));
 			return epoch;
 		}
 
-		public void SetLastEpoch(EpochRecord epoch) {
-			Ensure.NotNull(epoch, "epoch");
-
-			lock (_locker) {
-				if (epoch.EpochPosition > _lastEpochPosition) {
-					UpdateLastEpoch(epoch, flushWriter: false);
-					return;
-				}
-			}
-
+		public void CacheEpoch(EpochRecord epoch) {
+			var added = AddEpochToCache(epoch);
 			// Epoch record must have been already written, so we need to make sure it is where we expect it to be.
 			// If this check fails, then there is something very wrong with epochs, data corruption is possible.
-			if (!IsCorrectEpochAt(epoch.EpochPosition, epoch.EpochNumber, epoch.EpochId)) {
-				throw new Exception(string.Format("Not found epoch at {0} with epoch number: {1} and epoch ID: {2}. "
-				                                  + "SetLastEpoch FAILED! Data corruption risk!",
-					epoch.EpochPosition,
-					epoch.EpochNumber,
-					epoch.EpochId));
+			if (added && !IsCorrectEpochAt(epoch.EpochPosition, epoch.EpochNumber, epoch.EpochId)) {
```

Comment:
```suggestion
			if (!added && !IsCorrectEpochAt(epoch.EpochPosition, epoch.EpochNumber, epoch.EpochId)) {
```

@condron should it be `!added` ? the previous code was doing this check if it was not added to the cache.

Reply:
no, this should check only if this is a new item we did not write in this epoch manager.